File: apps/server/app/schemas/market.py
"""T2.3 市场 API schema —— **公开投影**（设计 §6）。

字段集即公开契约：schema 里没有的字段就是不对外暴露的（credential/runtime/provider/
memory/messages/drive/知识正文/审计 hash 等一律不在）。`owner_company_id` 与内部
`person_id` 同样不出现 —— 市场侧用 `listing_id` 与 `identity_id` 作为锚点。
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel

from app.schemas.competency import EmployeeCapabilitiesOut
from app.schemas.person import PersonKnowledgeSummaryOut


class MarketListingOut(BaseModel):
    """索引级挂牌投影（列表与详情头部共用）。"""

    listing_id: int
    identity_id: str
    name: str
    avatar: str = ""
    origin: str
    cultivation_state: str
    status: str
    quality_tier: str | None = None
    listed_at: datetime
    closed_at: datetime | None = None
    listed_by: str


class MarketListingItemOut(MarketListingOut):
    """列表项：在挂牌投影上追加 Fit 摘要（T2.5；仅请求带 position 时非空）。

    与挂牌响应（`MarketListingOut`）分开：POST 的返回形状是 T2.3 冻结的公开契约，
    不因 T2.5 的新字段而变化。
    """

    fit: MarketFitSummaryOut | None = None


class MarketListingPageOut(BaseModel):
    items: list[MarketListingItemOut]
    total: int
    limit: int
    offset: int


class MarketListingCreateIn(BaseModel):
    person_id: int
    #: 发行方档位（T2.4 起使用）；玩家挂牌留空
    quality_tier: str | None = None


class MarketTimelineEventOut(BaseModel):
    """履历事件公开投影：outcome 只含公开键（无 session_ids 等内部引用）。"""

    id: int
    kind: str
    topic: str
    outcome: dict
    occurred_at: datetime


class MarketEvidenceOut(BaseModel):
    """证据公开投影：保留 source_ref/assessment_run_id 以支持下钻追溯（验收 A）。"""

    id: int
    competency_code: str
    competency_name: str
    source_kind: str
    source_ref: str
    assessment_run_id: int | None = None
    signal: int | None = None
    quality: float | None = None
    occurred_at: datetime


class MarketFitEvaluationOut(BaseModel):
    """Fit 逐项评估的**公开投影**（无 requirement_id / competency_definition_id / inputs_hash）。"""

    code: str
    name: str
    domain_code: str
    kind: str
    requirement_type: str
    critical: bool
    minimum_score: int | None = None
    target_score: int | None = None
    minimum_confidence: float | None = None
    #: 候选人侧（null = 未评估，**不是 0**）
    candidate_score: int | None = None
    candidate_confidence: float | None = None
    evaluation_status: str
    reason_code: str
    gap_type: str | None = None
    is_unknown: bool
    is_strength: bool
    is_development_opportunity: bool
    margin_to_minimum: int | None = None
    margin_to_target: int | None = None


class MarketFitSummaryOut(BaseModel):
    """列表页的 Fit 摘要（详情页用 MarketFitOut）。"""

    position_definition_id: int
    fit_status: str
    qualification_status: str
    known_fit_score: float | None = None
    fit_confidence: float | None = None
    requirement_coverage: float
    known_count: int
    total_count: int


class MarketFitOut(BaseModel):
    """列表详情页的市场 Fit（public 投影：score/confidence/coverage/missing，无 inputs_hash）。"""

    listing_id: int
    position_definition_id: int
    position_code: str
    configured: bool
    profile_version_id: int | None = None
    profile_version: int | None = None
    fit_status: str
    qualification_status: str
    known_fit_score: float | None = None
    overall_fit_score: float | None = None
    fit_confidence: float | None = None
    requirement_coverage: float
    required_coverage: float
    preferred_coverage: float
    known_count: int
    total_count: int
    general_fit: float | None = None
    professional_fit: float | None = None
    strengths: list[MarketFitEvaluationOut]
    gaps: list[MarketFitEvaluationOut]
    uncertainties: list[MarketFitEvaluationOut]
    development_opportunities: list[MarketFitEvaluationOut]
    requirement_evaluations: list[MarketFitEvaluationOut]
    engine_version: str
    policy_version: str
    serializer_version: str
    calculated_at: datetime | None = None


class MarketListingSummaryOut(BaseModel):
    listing_id: int
    status: str
    quality_tier: str | None = None
    listed_at: datetime
    listed_by: str


class MarketCandidateIdentityOut(BaseModel):
    identity_id: str | None
    name: str
    avatar: str = ""
    origin: str | None = None
    cultivation_state: str | None = None


class MarketCandidateOut(BaseModel):
    """候选人公开档案：履历即证据链（score/confidence 并列，未评估如实 null）。"""

    listing: MarketListingSummaryOut
    identity: MarketCandidateIdentityOut
    traits: list[dict]
    competencies: EmployeeCapabilitiesOut
    knowledge_summary: PersonKnowledgeSummaryOut
    timeline: list[MarketTimelineEventOut]
    evidence: list[MarketEvidenceOut]
    market_state: str

File: apps/web/src/api/market.ts
import { del, get, post } from "./client";
import type { EmployeeCapabilities, TraitView } from "../types";
import type { EducationEvent } from "./cultivation";
import type { PersonKnowledgeSummary } from "./persons";

/**
 * T2.7 人才市场读面（docs/t2-talent-market-design.md §6/§8）。
 * 类型对齐 apps/server/app/schemas/market.py；改动后端 schema 时同步这里。
 *
 * 公开投影纪律：这里**没有** owner id / credential / 知识正文 —— 后端也不会返回。
 */

export interface MarketFitSummary {
  position_definition_id: number;
  fit_status: string;
  qualification_status: string;
  known_fit_score: number | null;
  fit_confidence: number | null;
  requirement_coverage: number;
  known_count: number;
  total_count: number;
}

export interface MarketListing {
  listing_id: number;
  identity_id: string;
  name: string;
  avatar: string;
  origin: string;
  cultivation_state: string;
  status: string;
  quality_tier: string | null;
  listed_at: string;
  closed_at: string | null;
  listed_by: string;
  /** 仅当检索带 position_definition_id 时非空（T2.5）。 */
  fit?: MarketFitSummary | null;
}

export interface MarketListingPage {
  items: MarketListing[];
  total: number;
  limit: number;
  offset: number;
}

export interface MarketFitEvaluation {
  code: string;
  name: string;
  domain_code: string;
  kind: string;
  requirement_type: string;
  critical: boolean;
  minimum_score: number | null;
  target_score: number | null;
  minimum_confidence: number | null;
  /** 候选人侧（null = 未评估，**不是 0**）。 */
  candidate_score: number | null;
  candidate_confidence: number | null;
  evaluation_status: string;
  reason_code: string;
  gap_type: string | null;
  is_unknown: boolean;
  is_strength: boolean;
  is_development_opportunity: boolean;
  margin_to_minimum: number | null;
  margin_to_target: number | null;
}

export interface MarketFit {
  listing_id: number;
  position_definition_id: number;
  position_code: string;
  configured: boolean;
  profile_version_id: number | null;
  profile_version: number | null;
  fit_status: string;
  qualification_status: string;
  known_fit_score: number | null;
  overall_fit_score: number | null;
  fit_confidence: number | null;
  requirement_coverage: number;
  required_coverage: number;
  preferred_coverage: number;
  known_count: number;
  total_count: number;
  general_fit: number | null;
  professional_fit: number | null;
  strengths: MarketFitEvaluation[];
  gaps: MarketFitEvaluation[];
  uncertainties: MarketFitEvaluation[];
  development_opportunities: MarketFitEvaluation[];
  requirement_evaluations: MarketFitEvaluation[];
  engine_version: string;
  policy_version: string;
  serializer_version: string;
  calculated_at: string | null;
}

export interface MarketEvidence {
  id: number;
  competency_code: string;
  competency_name: string;
  source_kind: string;
  source_ref: string;
  assessment_run_id: number | null;
  signal: number | null;
  quality: number | null;
  occurred_at: string;
}

export interface MarketCandidate {
  listing: {
    listing_id: number;
    status: string;
    quality_tier: string | null;
    listed_at: string;
    listed_by: string;
  };
  identity: {
    identity_id: string | null;
    name: string;
    avatar: string;
    origin: string | null;
    cultivation_state: string | null;
  };
  traits: TraitView[];
  competencies: EmployeeCapabilities;
  knowledge_summary: PersonKnowledgeSummary;
  timeline: EducationEvent[];
  evidence: MarketEvidence[];
  market_state: string;
}

export interface MarketListingFilters {
  text?: string;
  origin?: string;
  qualityTier?: string;
  /** 带职位时后端附 Fit 摘要并按匹配度排序（不筛人）。 */
  positionDefinitionId?: number | null;
  /** 只看本公司挂牌（T2.7a 的"我的挂牌"）。 */
  mine?: boolean;
  limit?: number;
  offset?: number;
}

export interface RecruitInput {
  department_id?: number | null;
  position_slot_id?: number | null;
  title?: string | null;
  role?: string | null;
  reason?: string;
}

export interface RecruitResult {
  person_id: number;
  employee_id: number;
  employee_slug: string;
  identity_id: string | null;
  company_id: number;
  listing_id: number;
  position_slot_id: number | null;
  assignment_id: number | null;
}

export function getMarketListings(filters: MarketListingFilters = {}): Promise<MarketListingPage> {
  const params = new URLSearchParams();
  if (filters.text) params.set("text", filters.text);
  if (filters.origin) params.set("origin", filters.origin);
  if (filters.qualityTier) params.set("quality_tier", filters.qualityTier);
  if (filters.positionDefinitionId != null) {
    params.set("position_definition_id", String(filters.positionDefinitionId));
  }
  if (filters.mine) params.set("mine", "true");
  if (filters.limit != null) params.set("limit", String(filters.limit));
  if (filters.offset != null) params.set("offset", String(filters.offset));
  const query = params.size > 0 ? `?${params.toString()}` : "";
  return get<MarketListingPage>(`/market/listings${query}`);
}

export function getMarketListing(listingId: number): Promise<MarketCandidate> {
  return get<MarketCandidate>(`/market/listings/${listingId}`);
}

export function getMarketFit(
  listingId: number,
  positionDefinitionId: number,
  profileVersionId?: number | null,
): Promise<MarketFit> {
  const params = new URLSearchParams({ position_definition_id: String(positionDefinitionId) });
  if (profileVersionId != null) params.set("profile_version_id", String(profileVersionId));
  return get<MarketFit>(`/market/listings/${listingId}/fit?${params.toString()}`);
}

/** 挂牌自己的角色（T2.3；幂等：已挂牌返回既有挂牌）。 */
export function createMarketListing(
  personId: number,
  qualityTier?: string | null,
): Promise<MarketListing> {
  return post<MarketListing>("/market/listings", {
    person_id: personId,
    ...(qualityTier ? { quality_tier: qualityTier } : {}),
  });
}

export function delistMarketListing(listingId: number): Promise<void> {
  return del<void>(`/market/listings/${listingId}`);
}

/** 招募（T2.6）：把既有 Person 变成本公司员工（不复制人级资产）。 */
export function recruitMarketListing(
  listingId: number,
  body: RecruitInput = {},
): Promise<RecruitResult> {
  return post<RecruitResult>(`/market/listings/${listingId}/recruit`, body);
}
